--- Django-1.0.2-final/django/forms/models.py.orig	2008-11-18 21:44:20.000000000 -0800
+++ Django-1.0.2-final/django/forms/models.py	2009-07-16 19:28:39.000000000 -0700
@@ -10,7 +10,7 @@
 
 from util import ValidationError, ErrorList
 from forms import BaseForm, get_declared_fields
-from fields import Field, ChoiceField, IntegerField, EMPTY_VALUES
+from fields import Field, ChoiceField, CharField, IntegerField, EMPTY_VALUES
 from widgets import Select, SelectMultiple, HiddenInput, MultipleHiddenInput
 from widgets import media_property
 from formsets import BaseFormSet, formset_factory, DELETION_FIELD_NAME
@@ -429,9 +429,13 @@
     def add_fields(self, form, index):
         """Add a hidden field for the object's primary key."""
         from django.db.models import AutoField
+        import ervin.noid
+
         self._pk_field = pk = self.model._meta.pk
         if pk.auto_created or isinstance(pk, AutoField):
             form.fields[self._pk_field.name] = IntegerField(required=False, widget=HiddenInput)
+        elif isinstance(pk, ervin.noid.NoidField):
+            form.fields[self._pk_field.name] = CharField(required=False, widget=HiddenInput)
         super(BaseModelFormSet, self).add_fields(form, index)
 
 def modelformset_factory(model, form=ModelForm, formfield_callback=lambda f: f.formfield(),
